File: backend/data/trigrams.py
# ─────────────────────────────────────────────────────────────
# The eight trigrams (八卦), keyed by 先天八卦数 (1–8).
# `lines` are bottom → top; True = yang (solid line).
#
# Ported from src/data/trigrams.ts — keep in sync.
# ─────────────────────────────────────────────────────────────
from typing import List

class Trigram:
    def __init__(self, no: int, pinyin: str, name: str, nature: str, symbol: str, lines: List[bool]):
        self.no = no
        self.pinyin = pinyin
        self.name = name
        self.nature = nature
        self.symbol = symbol
        self.lines = lines  # bottom → top

    def __repr__(self) -> str:
        return f"Trigram({self.no}, {self.name})"


TRIGRAMS = {
    1: Trigram(1, "Qián",  "The Creative",    "Heaven",  "☰", [True,  True,  True]),
    2: Trigram(2, "Duì",   "The Joyous",      "Lake",    "☱", [True,  True,  False]),
    3: Trigram(3, "Lí",    "The Clinging",    "Fire",    "☲", [True,  False, True]),
    4: Trigram(4, "Zhèn",  "The Arousing",    "Thunder", "☳", [True,  False, False]),
    5: Trigram(5, "Xùn",   "The Gentle",      "Wind",    "☴", [False, True,  True]),
    6: Trigram(6, "Kǎn",   "The Abysmal",     "Water",   "☵", [False, True,  False]),
    7: Trigram(7, "Gèn",   "Keeping Still",   "Mountain","☶", [False, False, True]),
    8: Trigram(8, "Kūn",   "The Receptive",   "Earth",   "☷", [False, False, False]),
}


def trigram_no_from_lines(lines: List[bool]) -> int:
    """Recover the 先天数 from three lines (bottom → top)."""
    v = (1 if lines[0] else 0) + (2 if lines[1] else 0) + (4 if lines[2] else 0)
    return 8 - v

File: src/data/trigrams.ts
// ─────────────────────────────────────────────────────────────
// The eight trigrams (八卦), keyed by 先天八卦数 (1–8).
// lines are bottom → top; true = yang (solid line).
// ─────────────────────────────────────────────────────────────

export interface Trigram {
  no: number
  pinyin: string
  name: string
  nature: string
  symbol: string
  lines: [boolean, boolean, boolean]
}

export const TRIGRAMS: Record<number, Trigram> = {
  1: { no: 1, pinyin: 'Qián', name: 'The Creative', nature: 'Heaven', symbol: '☰', lines: [true, true, true] },
  2: { no: 2, pinyin: 'Duì', name: 'The Joyous', nature: 'Lake', symbol: '☱', lines: [true, true, false] },
  3: { no: 3, pinyin: 'Lí', name: 'The Clinging', nature: 'Fire', symbol: '☲', lines: [true, false, true] },
  4: { no: 4, pinyin: 'Zhèn', name: 'The Arousing', nature: 'Thunder', symbol: '☳', lines: [true, false, false] },
  5: { no: 5, pinyin: 'Xùn', name: 'The Gentle', nature: 'Wind', symbol: '☴', lines: [false, true, true] },
  6: { no: 6, pinyin: 'Kǎn', name: 'The Abysmal', nature: 'Water', symbol: '☵', lines: [false, true, false] },
  7: { no: 7, pinyin: 'Gèn', name: 'Keeping Still', nature: 'Mountain', symbol: '☶', lines: [false, false, true] },
  8: { no: 8, pinyin: 'Kūn', name: 'The Receptive', nature: 'Earth', symbol: '☷', lines: [false, false, false] },
}

/** Recover the 先天数 from three lines (bottom → top). */
export function trigramNoFromLines(lines: boolean[]): number {
  const v = (lines[0] ? 1 : 0) + (lines[1] ? 2 : 0) + (lines[2] ? 4 : 0)
  return 8 - v
}
